Common/src/grid_movement_structure.cpp

```diff
@@ -8441,40 +8457,144 @@ void CFreeFormDefBox::SetSphe2Cart_CornerPoints(CConfig *config) {
+  int FFDBox_cgns_file;
+  int cell_dim, phys_dim;
+  int cgns_base, cgns_family, cgns_zone, cgns_err, dummy;
+  const char * basename;
+
+  /*--- FFD output is always 3D (even in 2D problems),
+   this is important for debuging ---*/
+  nDim = geometry->GetnDim();
+  cell_dim = nDim,
+  phys_dim = 3;
+
+  SPRINTF (FFDBox_filename, "ffd_boxes.cgns");
+
+  if ((original) && (iFFDBox == 0)) new_file = true;
+  else new_file = false;
+  
+  if (new_file) {
+    cgns_err = cg_open(FFDBox_filename, CG_MODE_WRITE, &FFDBox_cgns_file);
+    if (cgns_err) cg_error_print();
+    cgns_err = cg_descriptor_write("Title", "Visualization of the FFD boxes generated by SU2_DEF." );
+    if (cgns_err) cg_error_print();
+  }
+  else {
+    cgns_err = cg_open(FFDBox_filename, CG_MODE_MODIFY, &FFDBox_cgns_file);
+    if (cgns_err) cg_error_print();
+  }
+
+  if (original) {
+    basename = "Original_FFD";
+  }
+  else {
+    basename = "Deformed_FFD";
+  }
+
+  if (iFFDBox == 0){
+    cgns_err = cg_base_write(FFDBox_cgns_file, basename, cell_dim, phys_dim, &cgns_base);
+    if (cgns_err) cg_error_print();
+  }
+  cgns_err = cg_family_write(FFDBox_cgns_file, cgns_base, Tag.c_str(), &cgns_family);
+  if (cgns_err) cg_error_print();
+  
+  cgsize_t dims[9];
+  dims[0] = lDegree+1;
+  dims[1] = mDegree+1;
+  if (cell_dim == 3){
+     dims[2] = nDegree+1;
+  }
+  cgsize_t pointlen = 1;
+  for(int ii=0; ii<cell_dim; ii++){
+    dims[ii+cell_dim] = dims[ii] -1;
+    dims[ii+2*cell_dim] = 0;
+    pointlen *= dims[ii];
+  }
+
+  double *buffer = new double[pointlen];
+  SPRINTF (zonename, "SU2_Zone_%d", SU2_TYPE::Int(iFFDBox));
+  
+  cgns_err = cg_zone_write(FFDBox_cgns_file, cgns_base, zonename, dims, CGNS_ENUMV(Structured), &cgns_zone);
+  if (cgns_err) cg_error_print();
+  cgns_err = cg_goto(FFDBox_cgns_file, cgns_base, zonename, 0, NULL);
+  if (cgns_err) cg_error_print();
+  cgns_err = cg_famname_write(Tag.c_str());
+  if (cgns_err) cg_error_print();
+
+
+  const char* coord_names[3] = { "CoordinateX", "CoordinateY", "CoordinateZ" };
+  for (iDim=0; iDim<nDim; iDim++)
+  {
+    outDim = nDim == 2 ? 0: nDegree;
+    for (kDegree = 0; kDegree <= outDim; kDegree++) {
+      for (jDegree = 0; jDegree <= mDegree; jDegree++) {
+        for (iDegree = 0; iDegree <= lDegree; iDegree++) {
+          pos = iDegree + jDegree*(lDegree+1)+ kDegree*(lDegree+1)*(mDegree+1);
+          buffer[pos] = Coord_Control_Points[iDegree][jDegree][kDegree][iDim];
+        }
+      }
+    }
+    cgns_err = cg_coord_write(FFDBox_cgns_file, cgns_base, cgns_zone, CGNS_ENUMV(RealDouble), coord_names[iDim], buffer, &dummy);
+    if (cgns_err) cg_error_print();
+
+  }
+  if (nDim==2){
+    std::fill_n(buffer, pointlen, 0.0);
+    cgns_err = cg_coord_write(FFDBox_cgns_file, cgns_base, cgns_zone, CGNS_ENUMV(RealDouble), coord_names[2], buffer, &dummy);
+    if (cgns_err) cg_error_print();
+  }
+
+  delete [] buffer;
+  cgns_err = cg_close(FFDBox_cgns_file);
+  if (cgns_err) cg_error_print();
+#endif
```

Comment:
Maybe throw an error if the code was not compiled with CGNS support?

Reply:
Thanks for the review, I added a message as it is done elsewhere in the code.

Reply:
Thanks. I am not sure if this area of the code is executed in parallel, if it is, the message should only be printed by the MASTER_NODE, see e.g. line 211 of this file.

Reply:
This area of the code is only called by MASTER_NODE and thus not parallel see line 2739 of the file.

Reply:
Cool, thanks.